### PR TITLE
[Diloco] add DCN bandwidth throttling options

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

This change introduces programmatic traffic control ( tc ) bandwidth throttling support in MaxText and includes a dedicated DCN bandwidth microbenchmark script. This allows simulated network environment testing (e.g. limiting inter-slice DCN bandwidth) during multi-slice training.

# Key Changes

  • Centralized Throttling Utilities: Implemented  apply_dcn_throttling  and  cleanup_dcn_throttling  inside train_utils.py using the Linux traffic control ( tc ) utility with a Token Bucket Filter ( tbf ) queuing discipline.
  • Training Integration: Integrated DCN throttling within train.py's main training loop, applying configured traffic rules at startup and cleaning them up in the  finally  block of  train_loop .
  • Config Specifications: Introduced default configurations in base.yml and types.py for the following variables:

      -       dcn_bandwidth_limit
      -       dcn_bandwidth_burst
      -       dcn_bandwidth_latency
      -       dcn_bandwidth_interface

  • Microbenchmarking Script: Added  scripts/dcn_bandwidth_test.py  which creates a hybrid multi-slice device mesh and runs collective shard-mapped  psum  operations to measure and print latency/achieved DCN bandwidth under throttling constraints.
  

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/526650570

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Added tests in b/526650570

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: scripts/dcn_bandwidth_test.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+from maxtext.utils import train_utils
+
+
+def get_default_interface():
+  try:
+    route_output = subprocess.check_output("ip route show", shell=True, text=True)
+    for line in route_output.splitlines():
+      if "default" in line:
+        return line.split("dev")[1].strip().split()[0]
+  except Exception:  # pylint: disable=bare-except
```
Comment (richjames0) [inline scripts/dcn_bandwidth_test.py]: seems a bit broad

File: src/maxtext/configs/base.yml [hunk 2]
```diff
@@ -872,6 +872,10 @@ enable_diloco: false
 diloco_sync_period: 36
 diloco_outer_lr: 0.3
 diloco_outer_momentum: 0.9
+dcn_bandwidth_limit: ""
```
Comment (shralex) [inline src/maxtext/configs/base.yml:877]: Please add comments

File: scripts/dcn_bandwidth_test.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  return average_time_ms
+
+def create_mesh(dcn_size: int, ici_size: int):
+  """Creates a hybrid mesh with DCN and ICI axes."""
+  dcn_parallelism = [dcn_size, 1]
+  ici_parallelism = [1, ici_size]
+
+  total_devices = jax.device_count()
+  if total_devices != (dcn_size * ici_size):
+    raise ValueError(
```
Comment (richjames0) [inline scripts/dcn_bandwidth_test.py]: unnec. line break? or maybe it's right on the lien

File: scripts/dcn_bandwidth_test.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+# Copyright 2026 Google LLC
```
Comment (NuojCheng) [inline scripts/dcn_bandwidth_test.py]: why it is not under tests/.?

**Review by richjames0 — APPROVED** [review]
lgtm with a couple of nits

File: scripts/dcn_bandwidth_test.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  matrix_dims = [4096, 8192, 16384]
+  dtype = jnp.bfloat16
+  
+  interface = get_default_interface()
+  config = SimpleNamespace(
+      dcn_bandwidth_limit="28gbit",
+      dcn_bandwidth_burst="10mb",
+      dcn_bandwidth_latency="50ms",
+      dcn_bandwidth_interface=interface,
+  )
```
Comment (NuojCheng) [inline scripts/dcn_bandwidth_test.py]: maybe add more tests for different configs?

File: src/maxtext/trainers/pre_train/train.py [hunk 6]
```diff
@@ -607,6 +607,8 @@ def train_loop(config, recorder, state=None):
       state,
   ) = train_utils.setup_train_loop(config, recorder)
 
+  train_utils.apply_dcn_throttling(config)
```
Comment (NuojCheng) [inline src/maxtext/trainers/pre_train/train.py]: can we have a simple DCN performance regression test in PR description?

File: scripts/dcn_bandwidth_test.py (hunk 4 above)
Comment (khatwanimohit) [reply scripts/dcn_bandwidth_test.py]: This test is meant to run on multislice setup, I can add it there but then will have to --ignore in various places so pytest does not pick it up

File: src/maxtext/trainers/pre_train/train.py (hunk 6 above)
Comment (gobbleturk) [inline src/maxtext/trainers/pre_train/train.py]: rename method maybe_apply_dcn_throttling. Add a comment on this line that the default flag value is false, its otherwise concerning to see throttling applied in train.py

Comment (khatwanimohit) [reply src/maxtext/trainers/pre_train/train.py]: done

File: src/maxtext/configs/base.yml (hunk 2 above)
Comment (khatwanimohit) [reply src/maxtext/configs/base.yml:877]: done

File: scripts/dcn_bandwidth_test.py (hunk 4 above)
Comment (khatwanimohit) [reply scripts/dcn_bandwidth_test.py]: moved to tests

File: scripts/dcn_bandwidth_test.py (hunk 1 above)
Comment (khatwanimohit) [reply scripts/dcn_bandwidth_test.py]: fixed

File: scripts/dcn_bandwidth_test.py (hunk 3 above)
Comment (khatwanimohit) [reply scripts/dcn_bandwidth_test.py]: fixed

File: scripts/dcn_bandwidth_test.py (hunk 5 above)
Comment (khatwanimohit) [reply scripts/dcn_bandwidth_test.py]: Added tests for these cases 
```
all_cases = [
      ("none", "NO THROTTLING (Baseline)", False, None, None),
      ("100g", "100G Throttling", True, "100gbit", "600mb"),
      ("50g", "50G Throttling", True, "50gbit", "300mb"),
  ]
  ```

File: src/maxtext/trainers/pre_train/train.py (hunk 6 above)
Comment (khatwanimohit) [reply src/maxtext/trainers/pre_train/train.py]: added a bug link above

File: src/maxtext/utils/train_utils.py [hunk 7]
```diff
@@ -378,3 +380,50 @@ def validate_completed_steps(completed_steps: int, config_steps: int):
...
+        ["tc", "qdisc", "del", "dev", interface, "root"],
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+        check=False,
+    )
+  except Exception as e:  # pylint: disable=broad-exception-caught
+    max_logging.error(f"Failed to clean up existing traffic control on {interface}: {e}")
+
+  if not config.dcn_bandwidth_limit:
+    return
```
Comment (NuojCheng) [inline src/maxtext/utils/train_utils.py]: maybe we could move this if-else condition at the beginning

File: src/maxtext/utils/train_utils.py [hunk 8]
```diff
@@ -378,3 +380,50 @@ def validate_completed_steps(completed_steps: int, config_steps: int):
...
+  max_logging.log(f"Applying tc egress limit of {rate} (burst: {burst}, latency: {latency}) on {interface}...")
+  try:
+    cmd = ["tc", "qdisc", "add", "dev", interface, "root", "tbf", "rate", rate, "burst", burst, "latency", latency]
+    subprocess.run(cmd, check=True)
+    max_logging.log("DCN Bandwidth throttling applied successfully.")
+  except Exception as e:  # pylint: disable=broad-exception-caught
+    max_logging.error(f"Failed to apply DCN bandwidth throttling: {e}")
+
+
+def cleanup_dcn_throttling(config):
```
Comment (NuojCheng) [inline src/maxtext/utils/train_utils.py]: Could you set it as maybe_cleanup_dcn_throttling? The goal is that `if not config.dcn_bandwidth_limit` no additionaly logic/subprocess gets initiated in train.py.

File: src/maxtext/utils/train_utils.py (hunk 7 above)
Comment (khatwanimohit) [reply src/maxtext/utils/train_utils.py]: done

File: src/maxtext/utils/train_utils.py (hunk 8 above)
Comment (khatwanimohit) [reply src/maxtext/utils/train_utils.py]: done

**Review by NuojCheng — APPROVED** [review]
